spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java

```diff
@@ -29,24 +30,30 @@
 import org.springframework.util.StringUtils;
 
 /**
- * Utility for manipulating message headers related to span data.
+ * Creates a {@link Span.SpanBuilder} from {@link Message}
  *
- * @author Dave Syer
+ * @author Marcin Grzejszczak
  *
+ * @since 1.0.0
  */
-public class SpanMessageHeaders {
+class MessagingSpanInjector implements SpanInjector {
 
 	public static final String SPAN_HEADER = "X-Current-Span";
 
-	public static Span getSpanFromHeader(Message<?> message) {
-		if (message == null) {
-			return null;
-		}
-		Object object = message.getHeaders().get(SPAN_HEADER);
-		if (object instanceof Span) {
-			return (Span) object;
+	private final TraceKeys traceKeys;
+
+	public MessagingSpanInjector(TraceKeys traceKeys) {
+		this.traceKeys = traceKeys;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T> void inject(Span span, T carrier) {
```

Comment:
I think my other notes follow this, but this doesn't have to do with OT
unless the method is on the Tracer itself. The Tracer accepts variable
types as it is used in many contexts (ex messaging and also HTTP). This
class has a limited concern.

I would expect an implementation of OT to internally have inject adapters
for different carriers and the composite (tracer) would have a registry of
those adapters.

OT or not OT, methods are designed differently depending on the shape and
scope of the call site. It is a subtle, but important difference. I can
work with you on this today, and thanks for raising a straw man!